lib/domain.js

```diff
@@ -48,13 +48,61 @@ Object.defineProperty(process, 'domain', {
   }
 });
 
+let asyncHook = null;
+const pairing = new Map();
+function enableHook() {
+
+  if (asyncHook !== null) {
+    return;
+  }
+
+  asyncHook = async_hooks.createHook({
+    init(asyncId, type, triggerAsyncId, resource) {
+
+      if (process.domain !== null) { // if this operation is created while in a domain, let's mark it
+        pairing.set(asyncId, process.domain);
+        resource.domain = process.domain;
+        if (resource.promise !== undefined && resource.promise instanceof Promise) {
+          // resource.promise instanceof Promise make sure that the promise comes from the same context
+          // see https://github.com/nodejs/node/issues/15673
+          resource.promise.domain = process.domain;
+        }
+      }
+    },
+    before(asyncId) {
+
+      if (pairing.has(asyncId)) { // enter domain for this cb
+        process.domain = pairing.get(asyncId);
+      }
+    },
+    after(asyncId) {
+
+      if (pairing.has(asyncId)) { // exit domain for this cb
+        process.domain = null;
```

Comment:
Would it not be possible to exit one domain and enter another, as in nested domains?

Reply:
Ack. Would make sense yes!

Reply:
Actually, I should event use `domain.enter` and `domain.exit` for `before` and `after` hooks. That would use the domain stack.

Reply:
Absolutely!